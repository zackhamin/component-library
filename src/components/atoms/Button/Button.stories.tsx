import { Error } from 'Atoms/Icons';
import { Button, ButtonProps } from './Button';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'Atoms/Button',
	component: Button,
	children: {
		description: 'Text displayed within the button',
		control: {
			type: 'text',
		},
	},
} as Meta;

const defaultProps = {
	children: 'Button Text',
} as ButtonProps;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Secondary = Template.bind({});
Secondary.args = {
	...defaultProps,
	style: 'secondary',
};

export const Block = Template.bind({});
Block.args = {
	...defaultProps,
	block: true,
};

export const Text = Template.bind({});
Text.args = {
	...defaultProps,
	style: 'text',
	children: 'Text Button',
};

export const IconStory = Template.bind({});
IconStory.args = {
	style: 'icon',
	children: <Error color="#333333" fontSize="2.2rem" />,
};

export const SizeSm = Template.bind({});
SizeSm.args = {
	...defaultProps,
	size: 'small',
};

export const SizeMd = Template.bind({});
SizeMd.args = {
	...defaultProps,
	size: 'medium',
};

export const SizeLg = Template.bind({});
SizeLg.args = {
	...defaultProps,
	size: 'large',
};
