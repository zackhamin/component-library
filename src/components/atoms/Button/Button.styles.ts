import { colors } from 'Styles';
import styled from 'styled-components';

const SharedStyles = `
  background: ${colors.neonBlue};
  border-radius: 0.8rem;
  border: 2px solid ${colors.lightGrey};
  color: ${colors.white};
  cursor: pointer;
  font-weight: 700;
  min-width: 18rem;
  padding: 1.8rem 3.2rem;
  text-shadow: 0px 1px 1px rgba(0, 0, 0, 0.25);
  transition: 0.2s;
  white-space: nowrap;
	line-height: 1;

  &:hover, &:hover:not([disabled]) {
		background: ${colors.blue};
		text-decoration: none;
  }

  &.secondary {
		background: ${colors.white};
		border: 2px solid ${colors.darkGrey};
		border-radius: 6px;
		color: ${colors.darkGrey};
		text-shadow: none;
	}
	&.secondary:hover {
    background: ${colors.white} !important;
		border: 2px solid ${colors.black};
		color: ${colors.black};
		text-decoration: none;
	}

	&.icon {
		background: none;
		border: none;
		padding: 0.8rem;
		text-decoration: none;
		min-width: 0;
		text-shadow: none;

		&:hover {
			background: none;
		}

		&:focus {
			outline-color: ${colors.emerald}
		}
	}

  &.btn-medium {
    font-size: 1.6rem;
    min-width: 16rem;
    padding: 1.2rem 2.4rem;
  }

  &.btn-small {
		font-size: 1.4rem;
    min-width: 15rem;
		padding: 0.8rem 2.4rem;
	}

  &.block {
		display: block;
	}
`;

const StyledButton = styled.button`
	${SharedStyles}
	font-size: 1.6rem;
	display: flex;
	gap: 1rem;
	justify-content: center;

	&[disabled] {
		cursor: not-allowed !important;
		opacity: 0.5;
	}

	&.text {
		background: transparent;
		border: 0;
		border-radius: 0;
		border-bottom: 1px solid transparent;
		color: ${colors.grey};
		cursor: pointer;
		font-size: 1.6rem;
		font-weight: normal;
		min-width: auto;
		padding: 0;
		text-shadow: none;

		&:hover {
			background: transparent !important;
			border-bottom: 1px solid ${colors.darkGrey};
			color: ${colors.black};
		}

		&.btn-small {
			font-size: 1.4rem;
		}
		&.btn-medium {
			font-size: 1.6rem;
		}
		&.btn-large {
			font-size: 1.8rem;
		}
	}

	&.block {
		width: 100%;
	}
`;

export { StyledButton, SharedStyles };
