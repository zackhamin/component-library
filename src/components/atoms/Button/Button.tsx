import classNames from 'classnames';
import { forwardRef } from 'react';
import { StyledButton } from './Button.styles';

export interface ButtonProps
	extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'style'> {
	children: React.ReactNode;
	type?: 'submit' | 'button' | 'reset';
	style?: 'primary' | 'secondary' | 'icon' | 'text';
	size?: 'large' | 'medium' | 'small';
	block?: boolean;
	disabled?: boolean;
	onClick?: () => void;
	testId?: string;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
	(
		{
			children,
			type,
			style = 'primary',
			size = 'large',
			block = false,
			disabled = false,
			testId,
			onClick,
			...props
		},
		ref
	) => {
		const classes = classNames(style, `btn-${size}`, {
			block,
		});

		return (
			<StyledButton
				className={classes}
				onClick={onClick}
				type={type}
				disabled={disabled}
				data-testid={testId}
				ref={ref}
				{...props}
			>
				{children}
			</StyledButton>
		);
	}
);

export { Button };
