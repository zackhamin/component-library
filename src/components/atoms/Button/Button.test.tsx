import { Button } from './Button';
import userEvent from '@testing-library/user-event';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Button', () => {
	it('renders a button correctly', () => {
		const { container } = render(<Button onClick={jest.fn}>Click me!</Button>);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('calls the click handler', () => {
		const clickHandler = jest.fn();
		const { getByText } = render(
			<Button onClick={clickHandler}>Click me!</Button>
		);

		const button = getByText('Click me!');
		fireEvent.click(button);

		expect(clickHandler).toHaveBeenCalledTimes(1);
	});
});

describe('Disabled Button', () => {
	it('should have disabled attr when disabled prop set', () => {
		render(<Button disabled>Click me!</Button>);

		const button = screen.getByRole('button');

		expect(button).toHaveAttribute('disabled');
	});

	it('should not call onChange when disabled', () => {
		const clickHandler = jest.fn();
		render(
			<Button disabled onClick={clickHandler}>
				Click me!
			</Button>
		);

		const button = screen.getByRole('button');

		userEvent.click(button);

		expect(clickHandler).not.toHaveBeenCalled();
	});
});
