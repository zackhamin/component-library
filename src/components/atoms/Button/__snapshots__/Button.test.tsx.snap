// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a button correctly 1`] = `
.c0 {
  background: #1F51FF;
  border-radius: 0.8rem;
  border: 2px solid #EEEEEF;
  color: #ffffff;
  cursor: pointer;
  font-weight: 700;
  min-width: 18rem;
  padding: 1.8rem 3.2rem;
  text-shadow: 0px 1px 1px rgba(0,0,0,0.25);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  white-space: nowrap;
  line-height: 1;
  font-size: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover,
.c0:hover:not([disabled]) {
  background: #0000FF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.secondary {
  background: #ffffff;
  border: 2px solid #585858;
  border-radius: 6px;
  color: #585858;
  text-shadow: none;
}

.c0.secondary:hover {
  background: #ffffff !important;
  border: 2px solid #000000;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.icon {
  background: none;
  border: none;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-width: 0;
  text-shadow: none;
}

.c0.icon:hover {
  background: none;
}

.c0.icon:focus {
  outline-color: #00594C;
}

.c0.btn-medium {
  font-size: 1.6rem;
  min-width: 16rem;
  padding: 1.2rem 2.4rem;
}

.c0.btn-small {
  font-size: 1.4rem;
  min-width: 15rem;
  padding: 0.8rem 2.4rem;
}

.c0.block {
  display: block;
}

.c0[disabled] {
  cursor: not-allowed !important;
  opacity: 0.5;
}

.c0.text {
  background: transparent;
  border: 0;
  border-radius: 0;
  border-bottom: 1px solid transparent;
  color: #808080;
  cursor: pointer;
  font-size: 1.6rem;
  font-weight: normal;
  min-width: auto;
  padding: 0;
  text-shadow: none;
}

.c0.text:hover {
  background: transparent !important;
  border-bottom: 1px solid #585858;
  color: #000000;
}

.c0.text.btn-small {
  font-size: 1.4rem;
}

.c0.text.btn-medium {
  font-size: 1.6rem;
}

.c0.text.btn-large {
  font-size: 1.8rem;
}

.c0.block {
  width: 100%;
}

<button
  class="c0 primary btn-large"
>
  Click me!
</button>
`;
